packages/abi/src/parser/specifications/v1/resolvable-type.ts

```diff
@@ -0,0 +1,269 @@
+    abi: AbiSpecificationV1,
+    mt: AbiMetadataTypeV1,
+    typeArguments: AbiComponentV1['typeArguments']
+  ): ResolvableType | ResolvedType {
+    if (swayTypeMatchers.generic(mt.type)) {
+      const resolvedTypeParameter = parent.typeParamsArgsMap?.find(
+        ([tp]) => tp === mt.metadataTypeId
+      )?.[1];
+
+      return resolvedTypeParameter ?? new ResolvableType(abi, mt.metadataTypeId, undefined);
+    }
+
+    if (!mt.components) {
+      /**
+       * types like u8, u16 can make their way into metadata types
+       * if they aren't used _directly_ in a function-input/function-output/log/configurable/messageType
+       * These types are characterized by not having components and we can resolve them as-is
+       */
+      return new ResolvableType(abi, mt.metadataTypeId, undefined).resolveInternal(
+        mt.metadataTypeId,
+        undefined
+      );
+    }
+
+    const typeArgs = typeArguments?.map((ta) => this.handleComponent(parent, abi, ta).type);
+
+    const resolvable = new ResolvableType(
+      abi,
+      mt.metadataTypeId,
+      this.mapTypeParametersAndArgs(abi, mt, typeArgs)
+    );
+
+    if (typeArgs?.every((ta) => ta instanceof ResolvedType)) {
+      return resolvable.resolveInternal(mt.metadataTypeId, undefined);
+    }
+
+    if (resolvable.components?.every((comp) => comp.type instanceof ResolvedType)) {
+      return resolvable.resolveInternal(mt.metadataTypeId, undefined);
+    }
+
+    return resolvable;
+  }
+
+  private resolveInternal(
+    typeId: string | number,
+    typeParamsArgsMap: Array<[number, ResolvedType]> | undefined
+  ): ResolvedType {
+    const typeArgs = this.resolveTypeArgs(typeParamsArgsMap);
+
+    const components: ResolvedType['components'] = this.components?.map((c) => {
+      if (c.type instanceof ResolvedType) {
+        return c as { name: string; type: ResolvedType };
+      }
+
+      const resolvedGenericType = typeArgs?.find(
+        ([tp]) => (c.type as ResolvableType).metadataTypeId === tp
+      )?.[1];
+      if (resolvedGenericType) {
+        return {
+          name: c.name,
+          type: resolvedGenericType,
+        };
+      }
+      return { name: c.name, type: c.type.resolveInternal(c.type.metadataTypeId, typeArgs) };
+    });
+    return new ResolvedType({
+      type: this.metadataType.type,
+      typeId,
+      components,
+      typeParamsArgsMap: typeArgs,
+      metadataTypeId: this.metadataTypeId,
+    });
+  }
+
+  private resolveTypeArgs(
+    typeParamsArgsMap: Array<[number, ResolvedType]> | undefined
+  ): [number, ResolvedType][] | undefined {
+    return this.typeParamsArgsMap === undefined
+      ? typeParamsArgsMap
+      : this.typeParamsArgsMap.map(([typeParameter, value]) => {
+          if (value instanceof ResolvableType) {
+            const resolved = typeParamsArgsMap?.find(([tp]) => tp === value.metadataTypeId);
+
+            return (
+              resolved ?? [
+                +typeParameter,
```

Comment:
Why are you using a `+` here? Could `typeParameter` be null or a string instead of a number? It’s worth documenting this for clarity.